### DISCUSSION TITLE
@EntityGraph 와 fetchJoin의 차이점

### DISCUSSION THREAD

**Discussion by jimi567** [2024-woowacourse-study/level-interview#175] [Spring Data]
https://github.com/2024-woowacourse-study/level-interview/discussions/150

**Comment by J-I-H-O** [comment]
**<개요>**

`join fetch` 와 `@EntityGraph` 모두 연관된 엔티티를 함께 로딩할 때 사용하는 방법입니다.  N+1 문제를 해결하기 위해 사용할 수 있습니다.

**<차이점>**
- **원본 쿼리의 손상**
    join fetch 사용시 JPQL에 `join fetch` 구문이 추가된다는 단점이 존재합니다. 어떤 필드를 한번에 조회해야 한다는 사실까지 쿼리에서 표현하는 것은 불필요할 수 있습니다.
    반면 `@EntityGraph` 를 사용하면 원본 쿼리의 손상 없이 한번에 조회해올 필드를 지정해줄 수 있습니다.
    
    ```java
    // Academy+Subject+Teacher를 @EntityGraph 로 조회
    @EntityGraph(attributePaths = {"subjects", "subjects.teacher"})
    @Query("select a from Academy a")
    List<Academy> findAllEntityGraphWithTeacher();
    ```
    
- Inner Join vs Outer Join
    `join fetch` 는 Inner Join으로 동작하고, `@EntityGraph` 는 Outer Join으로 동작합니다.

https://jojoldu.tistory.com/165

**Comment by le2sky** [comment]
N + 1 문제를 해결하기 위해 `join fetch`와 `@EntityGraph`를 사용할 수 있습니다.
이를 사용하면 연관관계가 맺어진 엔티티를 한번에 가져와서 원치않은 추가 쿼리를 예방할 수 있습니다.
`join fetch`는 따로 명시하지 않는 경우 Inner Join, `@EntityGraph`는 Left Join이라는 차이점이 있습니다.

**Reply by le2sky** [reply]
+ 일반 join과 join fetch의 차이점?
